### PR TITLE
add url path for admission webhooks

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/kubernetes/kubernetes/issues/53826

Adds an optional field to admission webhook registration that allows a user to specify a path to post to.  This achieves parity with other webhooks.

Each segment is required to be a dns subdomain, which mirrors url rules for groups.

@kubernetes/api-reviewers 

```release-note
admission webhook registration now allows URL paths
```

### CODE REVIEW

File: pkg/apis/admissionregistration/validation/validation.go [hunk 1]
```diff
@@ -183,6 +183,16 @@ func validateExternalAdmissionHook(hook *admissionregistration.ExternalAdmission
 	if hook.FailurePolicy != nil && *hook.FailurePolicy != admissionregistration.Ignore {
 		allErrors = append(allErrors, field.NotSupported(fldPath.Child("failurePolicy"), *hook.FailurePolicy, []string{string(admissionregistration.Ignore)}))
 	}
+	if len(hook.URLPath) != 0 {
+		steps := strings.Split(hook.URLPath, "/")
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: `/foo/bar/` will validate `["","foo","bar",""]` and fail the first and last steps, right?

Comment (liggitt) [reply pkg/apis/admissionregistration/validation/validation.go]: if non-empty, must also start with a `/`

File: staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go [hunk 2]
```diff
@@ -157,6 +157,9 @@ type ExternalAdmissionHook struct {
 	// Required.
 	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
 
+	// URLPath is an optional field that specifies the URL path to use when posting the AdmissionReview object.
+	URLPath string `json:"urlPath" protobuf:"bytes,5,opt,name=urlPath"`
```
Comment (liggitt) [inline staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go]: belongs inside `AdmissionHookClientConfig`, right?

File: pkg/apis/admissionregistration/validation/validation.go (hunk 1 above)
Comment (deads2k) [reply pkg/apis/admissionregistration/validation/validation.go]: > if non-empty, must also start with a /

Ah, I was going the other way, no leading or trailing.  If you like though.  You want to force trailing or deny it?  I don't want to allow both.

File: staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go (hunk 2 above)
Comment (deads2k) [reply staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go]: > belongs inside AdmissionHookClientConfig, right?

Ok.

**Comment by deads2k** [comment]
comments addressed

File: pkg/apis/admissionregistration/validation/validation.go [hunk 3]
```diff
@@ -183,6 +183,24 @@ func validateExternalAdmissionHook(hook *admissionregistration.ExternalAdmission
 	if hook.FailurePolicy != nil && *hook.FailurePolicy != admissionregistration.Ignore {
 		allErrors = append(allErrors, field.NotSupported(fldPath.Child("failurePolicy"), *hook.FailurePolicy, []string{string(admissionregistration.Ignore)}))
 	}
+
+	if len(hook.ClientConfig.URLPath) != 0 {
+		if !strings.HasPrefix(hook.ClientConfig.URLPath, "/") || !strings.HasSuffix(hook.ClientConfig.URLPath, "/") {
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: requiring or forbidding trailing `/` is tricky, depending on the server implementing the webhook. why not allow either?

Comment (deads2k) [reply pkg/apis/admissionregistration/validation/validation.go]: > requiring or forbidding trailing / is tricky, depending on the server implementing the webhook. why not allow either?

done

File: pkg/apis/admissionregistration/validation/validation.go [hunk 4]
```diff
@@ -183,6 +183,29 @@ func validateExternalAdmissionHook(hook *admissionregistration.ExternalAdmission
...
+			allErrors = append(allErrors, field.Invalid(fldPath.Child("clientConfig", "urlPath"), hook.ClientConfig.URLPath, "must start with a '/'"))
+		}
+
+		urlPathToCheck := hook.ClientConfig.URLPath[1:]
+		if strings.HasSuffix(hook.ClientConfig.URLPath, "/") {
+			urlPathToCheck = urlPathToCheck[:len(urlPathToCheck)-1]
+		}
+		steps := strings.Split(urlPathToCheck, "/")
+		for i, step := range steps {
+			if len(step) == 0 {
```
Comment (liggitt) [inline pkg/apis/admissionregistration/validation/validation.go]: this breaks on `/`, right?

Comment (deads2k) [reply pkg/apis/admissionregistration/validation/validation.go]: > this breaks on /, right?

More tests added.

**Comment by liggitt** [comment]
cc @kubernetes/sig-api-machinery-bugs


/lgtm

**Comment by liggitt** [comment]
needs a release note

**Comment by deads2k** [comment]
only remaining approval is on docs which is autogenerated

**Comment by caesarxuchao** [comment]
lgtm. Thanks!

**Pull request merged**